src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveCompletionItemProvider.cs

```diff
@@ -58,14 +58,43 @@ internal static bool AtDirectiveCompletionPoint(RazorSyntaxTree syntaxTree, Sour
 
             // Do not provide IntelliSense for explicit expressions. Explicit expressions will usually look like:
             // [@] [(] [DateTime.Now] [)]
-            var isImplicitExpression = owner.FirstAncestorOrSelf<CSharpImplicitExpressionSyntax>() != null;
-            if (isImplicitExpression &&
-                owner.ChildNodes().All(n => n.IsToken && IsDirectiveCompletableToken((AspNetCore.Razor.Language.Syntax.SyntaxToken)n)))
+            var implicitExpression = owner.FirstAncestorOrSelf<CSharpImplicitExpressionSyntax>();
+            if (implicitExpression == null)
             {
-                return true;
+                return false;
+            }
+
+            if (owner.ChildNodes().Any(n => !n.IsToken || !IsDirectiveCompletableToken((AspNetCore.Razor.Language.Syntax.SyntaxToken)n)))
+            {
+                // Implicit expression contains invalid directive tokens
+                return false;
+            }
+
+            if (implicitExpression.FirstAncestorOrSelf<RazorDirectiveSyntax>() != null)
```

Comment:
Is there a reason why we're looking for specific *ancestors* instead of checking if there are *any* ancestors? Like, what's the top-level syntax node? should we check for that instead?

Reply:
I tried taking a current and future safe approach since I'm not 100% confident in the parent structure that exists in the various document kinds/locations. Some concerns I had in checking if `Parent.Parent` was the document node are:
- Does that change when close to statements or other constructs? Aka `@{ ... }Whappens here`. My concern being, "special" things occur when constructs are side by side on a single line and also we do things like reclassify whitespace after the end curly, will that have an impact?
- Will feature flags impact the structure? They already impact other scenarios so I don't want to revisit this code down the line if we change something that only impacts a certain feature flags flavor.

All that being said the opposite is also true where if we add a brand spanking new feature type that's not listed here directive attribute completion will show up. I just figured I'd rather break the new feature case than the we changed something in the SyntaxTree so don't ever show completions case.

Reply:
I agree. The top level will look like `RazorDocumentSyntax` -> `MarkupBlockSyntax` -> `CSharpBlockSyntax` -> `CSharpImplicitExpression`.

This just feels too fragile and might break if we add a new feature where we might want directive completion. What you have here seems reasonable.

Reply:
Ok. I don't love it but it seems like it's an improvment in general